flocker/provision/_gce.py

```diff
@@ -0,0 +1,494 @@
+
+        But, as that is untested, it will remain unimplemented.
+        """
+        raise NotImplementedError(
+            "GCE does not have reboot implemented because it has "
+            "experimentally been determined to not be needed.")
+
+
+@implementer(IProvisioner)
+class GCEProvisioner(PClass):
+    """
+    A provisioner that can create instances on GCE.
+
+    :ivar unicode zone: The zone in which instances will be provisioned in.
+    :ivar unicode project: The project under which instances will be
+        provisioned in.
+    :ivar Key ssh_public_key: The public ssh key that will transferred to the
+        instance for access.
+    :ivar compute: A Google Compute Engine Service that can be used to make
+        calls to the GCE API.
+    """
+
+    zone = field(type=unicode)
+    project = field(type=unicode)
+    ssh_public_key = field(type=Key)
+    compute = field()
+
+    def get_ssh_key(self):
+        return self.ssh_public_key
+
+    def create_node(self, name, distribution, metadata={}):
+        instance_name = _clean_to_gce_name(name)
+        ssh_key = unicode(self.ssh_public_key.toString('OPENSSH'))
+        config = _create_gce_instance_config(
+          instance_name=instance_name,
+          project=self.project,
+          zone=self.zone,
+          machine_type=_GCE_INSTANCE_TYPE,
+          image=_GCE_DISTRIBUTION_TO_IMAGE_MAP[distribution].get_active_image(
+              self.compute
+          )["selfLink"],
+          # The acceptance tests expect to be able to ssh in as root, but on
+          # CentOS that is by default turned off, so we must also enable sshing
+          # in as a different user to enable root ssh.
+          login_credentials=[
+              _LoginCredentials(
+                  username=_GCE_ACCEPTANCE_USERNAME,
+                  public_key=ssh_key
+              ),
+              _LoginCredentials(
+                  username=u"root",
+                  public_key=ssh_key
+              ),
+          ],
+          disk_size=_GCE_DISK_SIZE_GIB,
+          description=json.dumps({
+              u"description-format": u"v1",
+              u"created-by-python": u"flocker.provision._gce.GCEProvisioner",
+              u"name": name,
+              u"metadata": metadata
+          }),
+          tags=set([u"flocker-gce-provisioner",
+                    u"json-description",
+                    _GCE_FIREWALL_TAG]),
+          delete_disk_on_terminate=True,
+          startup_script=dedent("""\
+              #!/bin/sh
+              sed -i '/Defaults *requiretty/d' /etc/sudoers
+              """),
+        )
+
+        operation = self.compute.instances().insert(
+            project=self.project,
+            zone=self.zone,
+            body=config
+        ).execute()
+
+        operation_result = wait_for_operation(
+            self.compute, operation, timeout=60)
+
+        if not operation_result:
+            raise ValueError("Timed out waiting for VM creation")
```

Comment:
:question: it would be really nice if this gave a little more information about what went wrong.
Getting the current status would be a little tricky, but could it at least give the instance
id or something to make finding which one failed easier?